lxd/backup.go

```diff
@@ -353,7 +353,19 @@ func backupCreateTarball(s *state.State, path string, backup backup) error {
 	}
 
 	if compress != "none" {
-		compressedPath, err := compressFile(backupPath, compress)
+		infile, err := os.Open(backupPath)
+		if err != nil {
+			return err
+		}
+		defer infile.Close()
+
+		compressed, err := os.Create(backupPath + ".compressed")
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
+
+		err = compressFile(compress, infile, compressed)
```

Comment:
It seems you're using the same code in three places. :) Maybe a helper would be helpful (budam-tz).

Reply:
Sorry I didn't respond earlier.  I thought your whole comment was in jest.

This is an intentional change to move the opening of the input file and the creation of the output file out of compressFile so that finally in change number 5, the code in createImageFromContainer can pass in a reader which sends progress info.

I thought you were joking to say that I should undo this change.  In isolation, I would agree that it makes more sense to remove this duplication and put the Open/Create calls back into compressFile.

Do you have a specific suggestion of what sort of helper function could be added?

With the nature of the deferred calls to close, I think it is better to keep this code how it is.  But I could add something like:

func getFilesForCompress(path string) (os.File, so.File, string) {
  infile, err := os.Open(path)
  if err != nil {
    return nil, nil, err
  }
  outfile, err := os.Create(path + ".compressed")
  if err != nil {
    infile.Close()
    return nil, nil, err
  }
  return infile, outfile
}

Alternatively, I could leave the current API for compressFile how it is, and create a second version of compressFile which takes io.Reader/Writer.